Visualization - Scale to Data Bounds with a value

FilipRy opened the pull request:
## Summary

Closes #38536 

Checking the [Scale to Data Bound](https://www.elastic.co/guide/en/kibana/5.4/_metrics_amp_axes.html) option may cause that the bar, which value equals to the lower bounds is hidden. To avoid that, you can define bottom margin. Via bottom margin you specify a value, which lowers the lower bounds when displaying the plot.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [x] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- ~[ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~

### For maintainers

- ~[ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~
- [x] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

FilipRy commented:
@monfera could you pls review the changes here

elasticmachine commented:
Pinging @elastic/kibana-app

markov00 commented on x-pack/plugins/translations/translations/ja-JP.json:
```diff
@@ -2451,6 +2451,8 @@
     "kbnVislibVisTypes.controls.pointSeries.valueAxes.rotateLabel": "回転",
     "kbnVislibVisTypes.controls.pointSeries.valueAxes.scaleToDataBoundsLabel": "データバウンドに合わせる",
+    "kbnVislibVisTypes.controls.pointSeries.valueAxes.scaleToDataBounds.bottomMargin": "下余白",
+    "kbnVislibVisTypes.controls.pointSeries.valueAxes.scaleToDataBounds.minNeededBottomMargin": "下マージンは0以上でなければなりません",
```
It's not necessary to add the translation your self, those new ids are automatically extracted from the code before the next release, checked against existing and translated

FilipRy replied:
translation removed

markov00 commented on x-pack/plugins/translations/translations/zh-CN.json:
```diff
@@ -2029,6 +2029,8 @@
     "kbnVislibVisTypes.controls.pointSeries.valueAxes.rotateLabel": "旋转",
     "kbnVislibVisTypes.controls.pointSeries.valueAxes.scaleToDataBoundsLabel": "缩放到数据边界",
+    "kbnVislibVisTypes.controls.pointSeries.valueAxes.scaleToDataBounds.bottomMargin": "底部保证金",
+    "kbnVislibVisTypes.controls.pointSeries.valueAxes.scaleToDataBounds.minNeededBottomMargin": "底部保证金必须大于或等于0",
```
Same as comment before, it's not necessary to add the translation here, we will handle that before the release

FilipRy replied:
translation removed

markov00 commented on src/legacy/ui/public/vislib/lib/axis/axis_scale.js:
```diff
@@ -151,7 +151,15 @@ export class AxisScale {
+      if (scaleBottomMargin === 0) {
+        return domain;
+      } else {
+        domain[0] = domain[0] - scaleBottomMargin;
```
The `scaleBottomMargin` doesn't consider the case of a full negative set of data and a mixed negative/positive situation. It will be great if we can check that and adjust the domain accordingly.

FilipRy replied:
considering the negative and mixed (negative/positive evtl. with 0) -> see my comment below

markov00 commented on src/legacy/core_plugins/kbn_vislib_vis_types/public/controls/point_series/value_axes.html:
```diff
@@ -255,7 +255,33 @@
+                  class="kuiInput visEditorSidebar__input"
+                  type="number"
+                  step="0.1"
+                  greater-or-equal-than="0"
```
If we constraint here to only positive value, we should also configure how to handle the case of a visualization with all negative values

FilipRy commented:
@markov00  Thank you for the review!

I push commits to consider the cases of negative, zero and mixed values.
Basically I fixed that as follows:
- if there are only positive values -> add the margin to the bottom as before
- if there are only negative values -> add the margin to the top of the plot (i.e. plot's max = values' max + margin)
- if there are mixed values (non-negative and/or non-positive, positive and negative) -> ignore the margin, because no bar is hidden if scale to data bounds is checked.

Beside that, I renamed the bottomMargin to boundsMargin, I guess it's better name since it's not always a "bottom" margin anoymore.

Concerning the feature you suggested. I consider it interesting I can have a deeper look on that and start working.

markov00 commented:
jenkins test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2532/)

markov00 commented:
jenkins test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2755/)

markov00 commented:
Hey @FilipRy, we just merged your PR on master, we will backport it and it will be released as part of the 7.3 release.
Thanks again for the time dedicated on that feature. We really appreciated your efforts ❤️
